### DISCUSSION TITLE
Custom OAuth2 headers (especially Origin)

### DISCUSSION THREAD

**Discussion by robertburczak** [Kong/insomnia#3572] [Ideas]
Hi,

Is there a possibility to add custom headers to OAuth2 Authorization tab?

**Comment by orclev** [comment]
I just opened a feature request similar to this (#419) would that feature meet your need, or are you looking for a way of sending the token that doesn't use the Authorization header?

**Comment by robertburczak** [comment]
I am looking for possibility to send additional headers (name, value) as it is in Header tab for now, but in OAuth 2 tab.

**Comment by gschier** [comment]
@since78pl I'm curious, what kind of headers do you need to send with the OAuth requests?

**Comment by starJammer** [comment]
@gschier i'm guessing he's sending some custom headers that his authentication server listens to. Since the OAuth2 spec is pretty lax about additional things like this it isn't too surprising but I've not seen it done much.

**Comment by stale** [comment]
This issue has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.

**Comment by rupert-madden-abbott** [comment]
Can this please be looked at again?

Azure AD allows you to specify a "resource" header as part of the Access Token Request. This restricts the scope of your access token to the specific resources listed.

It doesn't seem to be possible to do this at the moment.

**Comment by JustinCoded** [comment]
Fitbit API also requires some credentials to be passed along in the Authorization header during exchange of auth code for access token, so this would be a great feature.

**Comment by p-m-j** [comment]
@DMarby accepting PRs for this or scheduling internally?

**Comment by DMarby** [comment]
@rustybox We'd be more then happy to accept a PR for it :)

**Comment by p-m-j** [comment]
Link to https://github.com/Kong/insomnia/issues/794 was lost on conversion from issue to discussion.

tl;dr lack of origin header prevents use of auth with PKCE flow when CORS is required for token endpoint which means one cannot use auth flow when Azure AD is the IdP.

Allowing for custom headers gives Azure AD users a workaround.

**Reply by wdawson** [reply]
Thanks for the re-link @rustybox . I came to the Insomnia team from Okta, so this makes sense to me. We're still happy to accept a PR if you're able. I'm not sure when we'll schedule this internally yet otherwise.

**Reply by p-m-j** [reply]
Newborn baby sapping a lot of my outside work hours, but I do plan to have a look at some point if no one else sorts it first.

**Reply by wdawson** [reply]
Congratulations on the family! And thanks in advance for taking a look.

**Reply by dimitropoulos** [reply]
Congrats to you and your fam @rustybox!

**Reply by p-m-j** [reply]
Many thanks :)

**Reply by develohpanda** [reply]
Congratulations!! 🎉
